### ISSUE TITLE
Getting internal compiler error

### ISSUE THREAD

**Issue by Alexnortung** [diku-dk/futhark#1650]
```
futhark test neural-network.fut
neural-network.fut:
Compiling with --backend=c:
Internal compiler error (unhandled IO exception).
Please report this at https://github.com/diku-dk/futhark/issues
BuilderT.lookupType: unknown variable output_m₁₈_9290
CallStack (from HasCallStack):
  error, called at src/Futhark/Builder.hs:107:18 in futhark-0.22.0-7LtyNk9C3pZ7TkI9jX4Ly4:Futhark.Builder
```

I am not really sure why I am getting this error. The code that produces this error is in https://github.com/Alexnortung/futhark-nn/tree/1315095645219c308efc3e6b753210c2455389ad
I am using the clean-ad branch of futhark.
The error occurs when running `futhark test neural-network.fut` from the `tests/` directory
But when I run `echo "[[[1.0,2.0,3.0,4.0,5.0],[1.0,2.0,3.0,4.0,5.0],[1.0,2.0,3.0,4.0,5.0],[1.0,2.0,3.0,4.0,5.0],[1.0,2.0,3.0,4.0,5.0], [1.0,2.0,3.0,4.0,5.0]]]" | futhark run -e nn_test neural-network.fut`

**Comment by Alexnortung** [comment]
It might have something to do with `neural-network/neural-netwok-new.fut:59`, when changing this line to `def conv_2d (m, n) = n_conv_2d m n`, it returns `Type variable output₂₀ cannot be instantiated with type containing anonymous sizes:`, which makes better sense

**Comment by athas** [comment]
Yeah, that's a bug in type inference.  I don't think this should be allowed, but I'll need to check carefully to be certain.

**Comment by athas** [comment]
Smaller program, same cause:

```
def f n m = \(b: bool) -> replicate n (replicate m b)

def g = uncurry f

def main x = map id (g x true)
```

**Comment by athas** [comment]
The type of `g` is actually sensible, but it looks like monomorphisation screws something up.

**Comment by athas** [comment]
It's not about `uncurry` or type inference at all:

```
def f : (i64, i64) -> ?[n][m].(b: bool) -> *[n][m]bool =
  \(n,m) (b: bool) -> replicate n (replicate m b)

def main x = map id (f x true)
```

I am not convinced we should allow these types.

**Comment by PhilipBorgesen** [comment]
I just encountered a similar issue when compiling `futhark-benchmarks/finpar/InterestCalib.fut`:
```
Internal compiler error (unhandled IO exception).
Please report this at https://github.com/diku-dk/futhark/issues
BuilderT.lookupType: unknown variable pre_106423
CallStack (from HasCallStack):
  error, called at src/Futhark/Builder.hs:107:18 in fthrk-0.21.9-b2c5b9db:Futhark.Builder
```

**Comment by athas** [comment]
I can compile InterestCalib just fine with the `c` backend; if there's an issue it's a different one.

**Comment by athas** [comment]
When this is fixed by #1650, your program will fail the type-check.  But the type error also links to an explanation how you can possibly fix it.  Not too often I get to conduct human trials!
